### DISCUSSION TITLE
Latest AIR SDK (33.1.1.758) - URLs to issues "AIR-XXXX" ?

### DISCUSSION THREAD

**Discussion by 2jfw** [airsdk/Adobe-Runtime-Support#1617] [Q&A] [answered]
@ajwfrost : Are there any public links to the issues mentioned in the latest AIR update **33.1.1.758** ?:

`AIR-5250: Removing HTMLLoader built-in webkit and using native StageWebView instead`
`AIR-5676: AS3 to support verbatim strings (similar to C#)`

I tried to find them in the **Adobe Bug Tracker** without success.
Also injecting the _IDs_ for the URL did not reveal them (eg. https://tracker.adobe.com/#/view/AIR-5250 or even just replacing last four digits of other AIR issues (391XXXX,419XXXX) like https://tracker.adobe.com/#/view/AIR-3915250 or https://tracker.adobe.com/#/view/AIR-4195250)

**Comment by ajwfrost** [comment, marked as answer]
Hi
No, the "AIR-xxx" refer to issues in our internal JIRA database. Mostly we use this github system and then in our JIRA tracker, we just link to these issues which we then put into the release notes. But there are a few things where we don't have an equivalent on here....

The "removing HTMLLoader" is a response to the problems that are being faced on macOS M1 hardware where the "WebKit" library that comes with AIR is not available for the M1 silicon. This means that apps which start up in arm64 mode and then try to use the HTMLLoader (or built-in StageWebView) will crash because the library loading fails. The WebKit library was 12 years old anyway and the effort of updating this was massive compared to the benefits that it might have brought: instead we are switching to the platform-native StageWebView implementations and also investigating whether third party browser engine solutions could be integrated.

For 'verbatim strings' this was just something that was fairly simple to do while we were working on another update of the AS3 compiler.. I get a bit bored of adding double-backslashes when copying/pasting a path to use for file input, hence removing the need to do this (i.e. the whole idea of escape characters) if you put an 'at' symbol in front of the string. So e.g.
```
   trace("This\that");
```
would end up displaying `This` and then a tab character and then `hat`... but
```
    trace(@"This\that");
```
really would display `This\that`

**Reply by 2jfw** [reply]
Thanks for the insight, Andrew! Much appreciated

**Reply by colm55** [reply]
Hi Andrew,
Is HTMLLoader being removed across all platforms or just MacOS?
Is any replacement for the SWF/HTML/JavaScript interaction being added to StageWebView?
Thanks,
Colm.

**Reply by ajwfrost** [reply]
Yes the WebKit library is being taken out on Windows too so the HTMLLoader class will for now just be a wrapper to the StageWebView 'native' implementation.

In terms of the interactions between the scripts, it won't be possible to do the native integration in the same way since what Adobe had done with WebKit was basically a large number of changes to the WebCore/JavaScriptCore libraries - and the amount of effort involved is one reason for that never having been updated... it would have needed to be redone every time WebKit changed...

But there should be ways of getting some integration working, and we may be able to help with setting some of this up using extra initialisation code, so if you can describe the sort of thing you were doing, we can take a look...

thanks

**Reply by colm55** [reply]
Hi Andrew,

Thanks for the VERY fast response! Much appreciated 👍 

The main SWF that I have is MASSIVE ... It's a maths assessment tool that I started about 20 years ago, so it's hard to say definitively what is in there right now! 😄 But ...

I currently have HTML radio buttons and text inputs that the AS3 reads the values from, and have HTML buttons that call AS3 code, and I use x, y, width, height values to align HTML and SWF objects.
I've used the .title/LOCATION_CHANGING StageWebView hack to exchange info in some parts of code, so can possibly use that more widely ... I haven't given it much thought as I only saw the AIR-5250 earlier today when I was checking out the latest AIR SDK.

This (ie removal of HTMLLoader) is a big loss of functionality IMO.
I'd guess that if yiz could even re-implement the old ExternalInterface-like calls on both sides that would at least provide a general method of exchange.

Thanks,
Colm.

**Reply by ajwfrost** [reply]
@colm55 can we discuss this further off-line? We can see what we can do then to mitigate this, it may be that we can tweak the APIs to make it simpler for things to work a bit like they did previously. ExternalInterface was only for the Flash Player (!) so I'm hoping you weren't actually loading content into the plug-in ... if you were, then you'd mostly be stuck on AIR SDK 19 from my understanding. But we can do similar sorts of APIs and then have some way to communicate between the AS3 and JavaScript/DOM maybe via an intermediate module...  please email me (andrew dot frost at harman dot com - whether that will stop spambots from finding my email, I doubt) and we can look into it!

thanks

**Reply by colm55** [reply]
Hi Andrew,
Email sent, but for record here:
For my money I think just allowing similar functionality to ExternalInterface-like function calling back and forth should cover most people’s needs, and keeping general DOM interaction within the webkit, and SWF interactions within AS3. Maybe a sandBoxBridge??
Thanks,
Colm.
